### PR TITLE
feat: defmodule snippet infer module name

### PR DESCRIPTION
**Pull request by lucacervello**
defmodule snippet infer module name

recording 🎥 

https://github.com/elixir-tools/next-ls/assets/5053083/e5e6aaa4-5a63-467b-b41f-49f6fd35a47a

### CODE REVIEW

File: lib/next_ls/snippet.ex [hunk 1]
```diff
@@ -171,7 +182,47 @@ defmodule NextLS.Snippet do
...
-  def get(_label, _trigger_character) do
+  def get(_label, _trigger_character, _opts) do
     nil
   end
+
+  defp infer_module_name(uri) do
+    result =
+      uri
+      |> Path.split()
+      |> Enum.reduce(false, fn
```
Comment (mhanberg) [inline lib/next_ls/snippet.ex:193]: Can you explain what the first element of the accumulator signifies here?

The clause with support is confusing me a tad.

**Comment by mhanberg** [comment]
You're on fire!

File: lib/next_ls/snippet.ex (hunk 1 above)
Comment (lucacervello) [reply lib/next_ls/snippet.ex:193]: Usually the module in `test/support` don't use `Test` in the module name like https://github.com/elixir-tools/next-ls/blob/main/test/support/utils.ex

> Can you explain what the first element of the accumulator signifies here?

do you mean `false`?

Comment (mhanberg) [reply lib/next_ls/snippet.ex:193]: I meant the tuple `{:test, []}`. I'm not sure what the atom is doing

Comment (lucacervello) [reply lib/next_ls/snippet.ex:193]: It's used [here](https://github.com/elixir-tools/next-ls/pull/398/files#diff-58f5082ef38a94ce607842d596a0ec1d974224124c9648642c2d46f5435ce750R200) to check if there is `support` under `test`

Comment (lucacervello) [reply lib/next_ls/snippet.ex:193]: Oh my bad, I'll try to explain.

I'm looking for three folders

- `lib` where usually the code is, `lib/foo.ex` it's translated to `Foo` (without `lib`)
- `test` where usually the tests are, `test/foo_test.exs` it's translated to `FooTest` (without `test`)
- `support` where usually are utils for tests, `test/support/foo.ex` it's translated to `Foo` (without `test` and `support`)

here are some tests that can help https://github.com/elixir-tools/next-ls/blob/9d65b759ac0cf86c943190342cf829097fc62346/test/next_ls/snippet_test.exs

Comment (mhanberg) [reply lib/next_ls/snippet.ex:193]: oh, i understand now, thanks!

basically you are just striping `lib/`, `test/` and `test/support/`  from the beginning for the path.

Comment (lucacervello) [reply lib/next_ls/snippet.ex:193]: Yes 👍

File: lib/next_ls/helpers/ast_helpers/env.ex [hunk 2]
```diff
@@ -113,6 +113,8 @@ defmodule NextLS.ASTHelpers.Env do
     }
   end
 
+  def ascend(nil, acc, _callback), do: acc
+
```
Comment (mhanberg) [inline lib/next_ls/helpers/ast_helpers/env.ex:117]: is this change necessary?

**Review by mhanberg — APPROVED** [review]
will merge after my last comment is answered.

File: lib/next_ls/helpers/ast_helpers/env.ex (hunk 2 above)
Comment (lucacervello) [reply lib/next_ls/helpers/ast_helpers/env.ex:117]: Yes

without it
```elixir
defmod|
```
don't trigger autocomplete because there is no previous node 

line here: https://github.com/elixir-tools/next-ls/pull/398/files/9d65b759ac0cf86c943190342cf829097fc62346#diff-934a3387a31dbe6cdfbab7475b2f2c24722ff71a7c11b282ef363ac6952f1c9fR19

File: lib/next_ls/snippet.ex (hunk 1 above)
Comment (scottming) [reply lib/next_ls/snippet.ex:193]: @lucacervello When you use someone else's code, please mention the source. If possible, contacting the author would be best; otherwise, it appears quite unethical.

https://github.com/lexical-lsp/lexical/blob/2178b4beab5d1a04cb83990e09b3dc232ed941cb/apps/server/lib/lexical/server/code_intelligence/completion/translations/macro.ex#L567C5-L603C8

Comment (mhanberg) [reply lib/next_ls/snippet.ex:193]: I've removed the code on main.

Sorry about that!
